commands/command_push.go

```diff
@@ -145,16 +145,30 @@ func lfsPushRefs(refnames []string, pushAll bool) ([]*git.RefUpdate, error) {
 		reflookup[ref.Name] = ref
 	}
 
+	suspiciousRefs := make([]string, 0)
+
 	refs := make([]*git.RefUpdate, len(refnames))
 	for i, name := range refnames {
 		if ref, ok := reflookup[name]; ok {
 			refs[i] = git.NewRefUpdate(cfg.Git, cfg.PushRemote(), ref, nil)
 		} else {
 			ref := &git.Ref{Name: name, Type: git.RefTypeOther, Sha: name}
+			suspiciousRefs = append(suspiciousRefs, name)
 			refs[i] = git.NewRefUpdate(cfg.Git, cfg.PushRemote(), ref, nil)
 		}
 	}
 
+	for _, suspiciousRef := range suspiciousRefs {
+		resolvedRef, err := git.ResolveRef(suspiciousRef)
+
+		if err != nil {
+			return nil, errors.New(tr.Tr.Get("Unable to determine ref validity of %v. Error: %v", suspiciousRef, err))
+		}
+		if resolvedRef == nil {
+			return nil, errors.New(tr.Tr.Get("Invalid ref: %v", suspiciousRef))
+		}
+	}
+
```

Comment:
This definitely resolves the problem and makes the new test in `t/t-push.sh` succeed, which is great (and the test fails without this change, so it's a good proof the fix works!)

Looking at the logic here more closely, I wonder if we could simply do this:
```patch
--- a/commands/command_push.go
+++ b/commands/command_push.go
@@ -157,6 +157,9 @@ func lfsPushRefs(refnames []string, pushAll bool) ([]*git.RefUpdate, error) {
 			refs[i] = git.NewRefUpdate(cfg.Git, cfg.PushRemote(), ref, nil)
 		} else {
 			ref := &git.Ref{Name: name, Type: git.RefTypeOther, Sha: name}
+			if _, err := git.ResolveRef(name); err != nil {
+				return nil, errors.New(tr.Tr.Get("Invalid ref argument: %v", name))
+			}
 			refs[i] = git.NewRefUpdate(cfg.Git, cfg.PushRemote(), ref, nil)
 		}
 	}
```

At first I thought we might replace the loop over the `suspiciousRefs` slice with a call to the `git.ResolveRefs()` [function](https://github.com/git-lfs/git-lfs/blob/baf40ac99850a62fe98515175d52df5c513463ec/git/git.go#L324-L335), which would do the loop for us.

Then I realized that since we're just going to report an error for the first ref which doesn't resolve, we don't really need to collect a list of them at all; we can just do the check right away while iterating over the ref names from the command line arguments.  So we can, I think, avoid the need to create `suspiciousRefs` in the first place.

Next I looked at [some](https://github.com/git-lfs/git-lfs/blob/baf40ac99850a62fe98515175d52df5c513463ec/commands/command_checkout.go#L101-L104) [other](https://github.com/git-lfs/git-lfs/blob/baf40ac99850a62fe98515175d52df5c513463ec/commands/command_migrate.go#L188-L191) [callers](https://github.com/git-lfs/git-lfs/blob/baf40ac99850a62fe98515175d52df5c513463ec/git/git.go#L327-L330) of `git.ResolveRef()` and concluded we don't need to check for a `nil` ref, just for a non-nil `err` return value.  That's because the function [checks](https://github.com/git-lfs/git-lfs/blob/baf40ac99850a62fe98515175d52df5c513463ec/git/git.go#L304-L306) for an empty response from `git rev-parse` and turns it into an error, and otherwise we can assume Git's response is valid enough to be parsed.

Finally, I think that if we use the same error message as the `git lfs fetch` command [generates](https://github.com/git-lfs/git-lfs/blob/baf40ac99850a62fe98515175d52df5c513463ec/commands/command_fetch.go#L52), we can ever-so-slightly simplify the job of a future contributor who's translating error messages for us, so they only have to translate one version of the same message rather than two.  This is, admittedly, a very minor concern.

Anyway, that's how I wound up at the suggestion above.  Would you be amenable to adopting that change?